core/cc_batch.go

```diff
@@ -84,64 +85,179 @@ func (cc *Chaincode) saveToBatch(
+
+	if len(result) != len(keys) {
+		return errors.New("len of result is not equal to len of keys")
 	}
 
-	defer func() {
-		err = stub.DelState(key)
+	err = cc.parsingResponseFromBatchKeys(batch, result)
+	if err != nil {
+		log.Errorf("couldn't parsing response from batch keys: %s", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (cc *Chaincode) collectKeysOfBatch(stub shim.ChaincodeStubInterface, batch *proto.Batch) ([]string, error) {
+	log := logger.Logger()
+
+	keys := make([]string, 0, len(batch.GetTxIDs())+len(batch.GetKeys())+len(batch.GetMultiSwapsKeys()))
+
+	for _, b := range batch.GetTxIDs() {
+		txID := hex.EncodeToString(b)
+		key, err := stub.CreateCompositeKey(config.BatchPrefix, []string{txID})
 		if err != nil {
-			log.Errorf("Couldn't delete from state tx %s: %s", txID, err.Error())
+			log.Errorf("Couldn't create composite key for tx %s: %s", txID, err.Error())
+			return nil, err
 		}
-	}()
+		keys = append(keys, key)
+	}
+
+	for _, sk := range batch.GetKeys() {
+		txID := hex.EncodeToString(sk.GetId())
+		key, err := stub.CreateCompositeKey(swap.SwapCompositeType, []string{txID})
+		if err != nil {
+			log.Errorf("Couldn't create composite key for tx %s: %s", txID, err.Error())
+			return nil, err
+		}
+		keys = append(keys, key)
+	}
+
+	for _, msk := range batch.GetMultiSwapsKeys() {
+		txID := hex.EncodeToString(msk.GetId())
+		key, err := stub.CreateCompositeKey(multiswap.MultiSwapCompositeType, []string{txID})
+		if err != nil {
+			log.Errorf("Couldn't create composite key for tx %s: %s", txID, err.Error())
+			return nil, err
+		}
+		keys = append(keys, key)
+	}
+
+	return keys, nil
+}
+
+func (cc *Chaincode) parsingResponseFromBatchKeys(batch *proto.Batch, result [][]byte) error {
+	log := logger.Logger()
+
+	batch.Pendings = make([]*proto.PendingTx, 0, len(batch.GetTxIDs()))
+	for i, b := range batch.GetTxIDs() {
+		if result[i] == nil {
+			batch.Pendings = append(batch.Pendings, nil)
+			continue
+		}
+		txID := hex.EncodeToString(b)
+		pending := new(proto.PendingTx)
+		if err := pb.Unmarshal(result[i], pending); err != nil {
+			log.Errorf("couldn't unmarshal transaction %s: %s", txID, err.Error())
+			return err
+		}
+		batch.Pendings = append(batch.Pendings, pending)
+	}
+
+	result = result[len(batch.GetTxIDs()):]
+
+	for i, sk := range batch.GetKeys() {
+		sks := &proto.SwapKey_Swap{}
+		batch.GetKeys()[i].Payload = sks
+
+		if result[i] == nil {
+			continue
+		}
+
+		txID := hex.EncodeToString(sk.GetId())
+		s := new(proto.Swap)
+		if err := pb.Unmarshal(result[i], s); err != nil {
+			log.Errorf("couldn't unmarshal transaction %s: %s", txID, err.Error())
+			return err
+		}
+
+		sks.Swap = s
+	}
+
+	result = result[len(batch.GetKeys()):]
+
+	for i, sk := range batch.GetMultiSwapsKeys() {
+		skms := &proto.SwapKey_MultiSwap{}
+		batch.GetMultiSwapsKeys()[i].Payload = skms
+
+		if result[i] == nil {
+			continue
+		}
+
+		txID := hex.EncodeToString(sk.GetId())
+		ms := new(proto.MultiSwap)
```

Comment:
inconsistent (mixed) styles

new(...) or &...{}

Reply:
that's legacy